common/caching/immutableinmemorylru.go

```diff
@@ -0,0 +1,47 @@
+package caching
+
+import (
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
+type ImmutableInMemoryLRUCache struct {
+	roomVersions *lru.Cache
+}
+
+func NewImmutableInMemoryLRUCache() (*ImmutableInMemoryLRUCache, error) {
+	roomVersionCache, rvErr := lru.New(RoomVersionMaxCacheEntries)
+	if rvErr != nil {
+		return nil, rvErr
+	}
+	return &ImmutableInMemoryLRUCache{
+		roomVersions: roomVersionCache,
+	}, nil
+}
+
+func checkForInvalidMutation(cache *lru.Cache, key string, value interface{}) {
+	if peek, ok := cache.Peek(key); ok && peek != value {
+		panic("invalid use of immutable cache tries to mutate existing value")
+	}
+}
+
+func (c *ImmutableInMemoryLRUCache) GetRoomVersion(roomID string) (gomatrixserverlib.RoomVersion, bool) {
+	if c == nil || !RoomVersionCachingEnabled {
```

Comment:
Right so this is where we do `nil` guards. I personally think it's a little premature to add options for enabling/disabling certain caches at this stage, as we don't yet have use cases for it (eg maybe doing it at compile time like this won't be what we want).

Reply:
I've removed the enable flags.